File: windows/src/FileTranscriber.cs
namespace Golos;

/// <summary>
/// Расшифровка готового файла: встречи, созвона, записи экрана.
///
/// Отличие от диктовки в том, что результат нужен с тайм-кодами и ложится
/// на диск. Всё остальное — нарезка на куски, переспрос без словаря, чистка
/// выдумок — то же самое.
/// </summary>
public sealed class FileTranscriber
{
    public enum Stage { Reading, Recognizing, Writing }

    public sealed record Result(string TextFile, string SubtitleFile, string? AudioFile,
                                int Segments, double Duration);

    private readonly Whisper whisper;
    private volatile bool cancelled;

    public FileTranscriber(Whisper whisper) => this.whisper = whisper;

    public void Cancel() => cancelled = true;

    public async Task<Result> TranscribeAsync(string path, Config config,
                                              Action<Stage, double> onProgress)
    {
        cancelled = false;
        onProgress(Stage.Reading, 0);

        var wav = await Task.Run(() => MediaDecoder.Decode(path, f => onProgress(Stage.Reading, f)));

        var chunks = AudioSplit.ChunksWithOffsets(wav);
        Log.Write($"расшифровка файла «{Path.GetFileName(path)}»: кусков {chunks.Count}");

        var segments = new List<Whisper.Segment>();
        for (var i = 0; i < chunks.Count; i++)
        {
            if (cancelled) throw new OperationCanceledException();
            onProgress(Stage.Recognizing, (double)i / chunks.Count);
            segments.AddRange(await AskChunkAsync(chunks[i], config.PromptString()));
        }

        onProgress(Stage.Writing, 1);
        return Write(segments, wav, path, config);
    }

    /// <summary>
    /// Один кусок с теми же защитами, что и диктовка: заученная фраза,
    /// зацикливание и пустота лечатся переспросом без словаря.
    /// </summary>
    private async Task<List<Whisper.Segment>> AskChunkAsync(AudioSplit.Chunk chunk, string prompt)
    {
        var raw = await whisper.TranscribeSegmentsAsync(chunk.Wav, prompt);
        var joined = raw == null ? "" : string.Join(" ", raw.Select(s => s.Text));
        var seconds = (chunk.Wav.Length - 44) / 32000.0;

        var suspicious = raw == null
            || joined.Length == 0
            || Hallucination.LooksInvented(joined, seconds)
            || Hallucination.LooksDegenerate(joined);

        if (suspicious && prompt.Length > 0)
        {
            Log.Write($"кусок не разобрался: «{Shorten(joined)}» — переспрашиваю без словаря");
            raw = await whisper.TranscribeSegmentsAsync(chunk.Wav, "");
        }

        if (raw == null) return new List<Whisper.Segment>();

        // Тайм-коды приходят от начала куска — сдвигаем к началу файла.
        return raw.Select(s => s with { Start = s.Start + chunk.StartSeconds,
                                        End = s.End + chunk.StartSeconds }).ToList();
    }

    private static string Shorten(string s) => s.Length > 60 ? s[..60] : s;

    private Result Write(List<Whisper.Segment> segments, byte[] wav, string source, Config config)
    {
        var folder = OutputFolder(source, config);
        var name = Path.GetFileNameWithoutExtension(source);
        var cleaned = segments
            .Select(s => s with { Text = config.ApplyReplacements(s.Text).Trim() })
            .Where(s => s.Text.Length > 0)
            .ToList();

        Directory.CreateDirectory(folder);

        var textFile = Path.Combine(folder, name + ".txt");
        File.WriteAllText(textFile, PlainText(cleaned));

        var subtitleFile = Path.Combine(folder, name + ".srt");
        File.WriteAllText(subtitleFile, Subtitles(cleaned));

        string? audioFile = null;
        if (config.KeepConvertedAudio)
        {
            audioFile = Path.Combine(folder, name + ".wav");
            File.WriteAllBytes(audioFile, wav);
        }

        var duration = cleaned.Count > 0 ? cleaned[^1].End : 0;
        Log.Write($"расшифровка готова: {textFile}, отрезков {cleaned.Count}");
        return new Result(textFile, subtitleFile, audioFile, cleaned.Count, duration);
    }

    /// <summary>Пусто в настройках — кладём рядом с исходником.</summary>
    private static string OutputFolder(string source, Config config)
    {
        var stored = config.FileOutputFolder.Trim();
        return stored.Length == 0 ? Path.GetDirectoryName(source)! : stored;
    }

    /// <summary>
    /// Текст абзацами: whisper режет речь на куски по несколько слов,
    /// и построчно читать это невозможно.
    /// </summary>
    private static string PlainText(List<Whisper.Segment> segments)
    {
        const int minimumLine = 90;
        const int maximumLine = 400;

        var lines = new List<string>();
        var buffer = "";
        double start = 0;

        foreach (var segment in segments)
        {
            if (buffer.Length == 0)
            {
                start = segment.Start;
                buffer = segment.Text;
            }
            else
            {
                // Разорванное слово склеиваем без пробела, иначе выходит
                // «стар ую версию».
                buffer += segment.JoinsPreviousWord ? segment.Text : " " + segment.Text;
            }

            var ended = ".!?…".Contains(buffer[^1]);
            if ((ended && buffer.Length >= minimumLine) || buffer.Length >= maximumLine)
            {
                lines.Add($"[{Timecode(start)}] {buffer}");
                buffer = "";
            }
        }
        if (buffer.Length > 0) lines.Add($"[{Timecode(start)}] {buffer}");
        return string.Join(Environment.NewLine + Environment.NewLine, lines);
    }

    private static string Subtitles(List<Whisper.Segment> segments)
    {
        var blocks = segments.Select((s, i) =>
            $"{i + 1}{Environment.NewLine}{SrtTime(s.Start)} --> {SrtTime(s.End)}{Environment.NewLine}{s.Text}");
        return string.Join(Environment.NewLine + Environment.NewLine, blocks) + Environment.NewLine;
    }

    private static string Timecode(double seconds)
    {
        var t = (int)seconds;
        return $"{t / 3600:D2}:{t % 3600 / 60:D2}:{t % 60:D2}";
    }

    private static string SrtTime(double seconds)
    {
        var t = (int)seconds;
        var ms = (int)((seconds - t) * 1000);
        return $"{t / 3600:D2}:{t % 3600 / 60:D2}:{t % 60:D2},{ms:D3}";
    }
}

File: Sources/Golos/FileTranscriber.swift
import AppKit
import Foundation

/// Расшифровка готового файла: встречи, созвона, записи экрана.
///
/// Отличие от диктовки в том, что результат нужен с тайм-кодами и ложится
/// на диск, а не в поле ввода. Всё остальное — нарезка на куски, переспрос
/// без словаря, чистка выдумок — то же самое.
@MainActor
final class FileTranscriber {
    enum Stage {
        case reading        // достаём звук из файла
        case recognizing    // гоняем куски через движок
        case writing
    }

    struct Progress {
        var stage: Stage
        var fraction: Double
    }

    struct Result {
        var textFile: URL
        var subtitleFile: URL
        var audioFile: URL?
        var segments: Int
        var duration: TimeInterval
    }

    private let whisper: Whisper
    private var cancelled = false

    init(whisper: Whisper) {
        self.whisper = whisper
    }

    func cancel() { cancelled = true }

    func transcribe(url: URL,
                    config: Config,
                    onProgress: @escaping (Progress) -> Void,
                    onFinish: @escaping (Swift.Result<Result, Error>) -> Void) {
        cancelled = false
        onProgress(Progress(stage: .reading, fraction: 0))

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self else { return }

            let wav: Data
            do {
                wav = try MediaDecoder.wav(from: url) { fraction in
                    DispatchQueue.main.async {
                        onProgress(Progress(stage: .reading, fraction: fraction))
                    }
                }
            } catch {
                DispatchQueue.main.async { onFinish(.failure(error)) }
                return
            }

            DispatchQueue.main.async {
                self.recognize(wav: wav, source: url, config: config,
                               onProgress: onProgress, onFinish: onFinish)
            }
        }
    }

    // MARK: - Распознавание

    private func recognize(wav: Data, source: URL, config: Config,
                           onProgress: @escaping (Progress) -> Void,
                           onFinish: @escaping (Swift.Result<Result, Error>) -> Void) {
        let chunks = AudioSplit.chunksWithOffsets(wav: wav)
        Log.write("расшифровка файла «\(source.lastPathComponent)»: кусков \(chunks.count)")

        var segments: [Whisper.Segment] = []
        var index = 0

        func step() {
            guard !cancelled else {
                onFinish(.failure(CancellationError()))
                return
            }
            guard index < chunks.count else {
                onProgress(Progress(stage: .writing, fraction: 1))
                finish(segments: segments, wav: wav, source: source,
                       config: config, onFinish: onFinish)
                return
            }

            let chunk = chunks[index]
            onProgress(Progress(stage: .recognizing,
                                fraction: Double(index) / Double(chunks.count)))

            askChunk(chunk, prompt: config.promptString, attempt: 1) { pieces in
                segments.append(contentsOf: pieces)
                index += 1
                step()
            }
        }
        step()
    }

    /// Один кусок с теми же защитами, что и диктовка: заученная фраза,
    /// зацикливание и пустота лечатся переспросом без словаря.
    private func askChunk(_ chunk: AudioSplit.Chunk, prompt: String, attempt: Int,
                          completion: @escaping ([Whisper.Segment]) -> Void) {
        whisper.transcribeSegments(wav: chunk.wav, prompt: prompt) { [weak self] result in
            guard let self else { return }
            guard case .success(let raw) = result else {
                completion([])
                return
            }

            let joined = raw.map(\.text).joined(separator: " ")
            let seconds = Double(chunk.wav.count - 44) / 32000
            let suspicious = Hallucination.looksInvented(text: joined, audioDuration: seconds)
                || Hallucination.looksDegenerate(joined)

            if attempt == 1, suspicious || joined.isEmpty, !prompt.isEmpty {
                Log.write("кусок не разобрался: «\(joined.prefix(60))» — переспрашиваю без словаря")
                self.askChunk(chunk, prompt: "", attempt: 2, completion: completion)
                return
            }

            // Тайм-коды приходят от начала куска — сдвигаем к началу файла.
            let shifted = raw.map {
                Whisper.Segment(start: $0.start + chunk.startSeconds,
                                end: $0.end + chunk.startSeconds,
                                text: $0.text,
                                joinsPreviousWord: $0.joinsPreviousWord)
            }
            completion(shifted)
        }
    }

    // MARK: - Запись результата

    private func finish(segments: [Whisper.Segment], wav: Data, source: URL, config: Config,
                        onFinish: @escaping (Swift.Result<Result, Error>) -> Void) {
        let base = source.deletingPathExtension().lastPathComponent
        let cleaned = clean(segments, config: config)

        // Папка рядом с исходником бывает недоступна для записи: диктофон,
        // флешка, сетевой диск, образ. Терять из-за этого сорок минут работы
        // нельзя, поэтому дальше спрашиваем, куда положить, а если человек
        // отказался — кладём в Загрузки, лишь бы не пропало.
        let preferred = outputFolder(for: source, config: config)
        var folders = [preferred]
        if !FileTranscriber.writable(preferred), let chosen = askWhereToSave(base: base, source: source) {
            folders.append(chosen)
        }
        if let downloads = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first {
            folders.append(downloads)
        }
        folders.append(Config.directory.appendingPathComponent("Расшифровки"))

        var lastError: Error?
        for (index, folder) in folders.enumerated() {
            do {
                try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)

                let textFile = folder.appendingPathComponent("\(base).txt")
                try plainText(cleaned).write(to: textFile, atomically: true, encoding: .utf8)

                let subtitleFile = folder.appendingPathComponent("\(base).srt")
                try subtitles(cleaned).write(to: subtitleFile, atomically: true, encoding: .utf8)

                var audioFile: URL?
                if config.keepConvertedAudio {
                    let target = folder.appendingPathComponent("\(base).wav")
                    try wav.write(to: target)
                    audioFile = target
                }

                if index > 0 {
                    Log.write("папка «\(folders[0].path)» недоступна для записи (\(lastError?.localizedDescription ?? "?")), расшифровка сохранена в \(folder.path)")
                }
                let duration = cleaned.last?.end ?? 0
                Log.write("расшифровка готова: \(textFile.path), отрезков \(cleaned.count)")
                onFinish(.success(Result(textFile: textFile, subtitleFile: subtitleFile,
                                         audioFile: audioFile, segments: cleaned.count,
                                         duration: duration)))
                return
            } catch {
                lastError = error
                // Мусор от половины записи убираем, чтобы рядом с исходником
                // не оставался обрезанный txt без субтитров.
                try? FileManager.default.removeItem(at: folder.appendingPathComponent("\(base).txt"))
            }
        }

        Log.write("расшифровку не удалось сохранить никуда: \(lastError?.localizedDescription ?? "?")")
        onFinish(.failure(lastError ?? DecodeFailure.nowhereToWrite))
    }

    enum DecodeFailure: LocalizedError {
        case nowhereToWrite
        var errorDescription: String? { "Некуда сохранить расшифровку" }
    }

    private func clean(_ segments: [Whisper.Segment], config: Config) -> [Whisper.Segment] {
        segments.compactMap { segment in
            let text = config.applyReplacements(to: segment.text)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            guard !text.isEmpty else { return nil }
            return Whisper.Segment(start: segment.start, end: segment.end, text: text,
                                   joinsPreviousWord: segment.joinsPreviousWord)
        }
    }

    /// Пусто в настройках — кладём рядом с исходником, как и договаривались.
    /// Можно ли писать в папку. Проверяем заранее, чтобы спросить человека
    /// до того, как он увидит ошибку, а не после.
    static func writable(_ folder: URL) -> Bool {
        let fm = FileManager.default
        if !fm.fileExists(atPath: folder.path) {
            // Папки ещё нет: создастся ли она, покажет попытка.
            guard (try? fm.createDirectory(at: folder, withIntermediateDirectories: true)) != nil
            else { return false }
        }
        return fm.isWritableFile(atPath: folder.path)
    }

    /// Спрашивает, куда положить расшифровку. Возвращает nil, если отказались.
    private func askWhereToSave(base: String, source: URL) -> URL? {
        let panel = NSSavePanel()
        panel.title = "Куда сохранить расшифровку"
        panel.message = "Рядом с «\(source.lastPathComponent)» записать нельзя: том только для чтения."
        panel.nameFieldStringValue = "\(base).txt"
        panel.canCreateDirectories = true
        panel.directoryURL = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first

        NSApp.activate(ignoringOtherApps: true)
        guard panel.runModal() == .OK, let url = panel.url else {
            Log.write("сохранение отменили, кладу в Загрузки")
            return nil
        }
        // Имя файла складывает сама запись: рядом лягут ещё .srt и, может, .wav.
        return url.deletingLastPathComponent()
    }

    private func outputFolder(for source: URL, config: Config) -> URL {
        let stored = config.fileOutputFolder.trimmingCharacters(in: .whitespaces)
        guard !stored.isEmpty else { return source.deletingLastPathComponent() }
        return URL(fileURLWithPath: (stored as NSString).expandingTildeInPath)
    }

    /// Текст абзацами: whisper режет речь на куски по несколько слов,
    /// и построчно читать это невозможно. Склеиваем до конца предложения,
    /// время берём от первого отрезка в абзаце.
    private func plainText(_ segments: [Whisper.Segment]) -> String {
        let minimumLine = 90
        let maximumLine = 400

        var lines: [String] = []
        var buffer = ""
        var start: Double = 0

        for segment in segments {
            if buffer.isEmpty { start = segment.start }
            if buffer.isEmpty {
                buffer = segment.text
            } else {
                // Разорванное слово склеиваем без пробела, иначе выходит
                // «стар ую версию».
                buffer += segment.joinsPreviousWord ? segment.text : " " + segment.text
            }

            let ended = buffer.last.map { ".!?…".contains($0) } ?? false
            if (ended && buffer.count >= minimumLine) || buffer.count >= maximumLine {
                lines.append("[\(timecode(start))] \(buffer)")
                buffer = ""
            }
        }
        if !buffer.isEmpty { lines.append("[\(timecode(start))] \(buffer)") }
        return lines.joined(separator: "\n\n")
    }

    private func subtitles(_ segments: [Whisper.Segment]) -> String {
        segments.enumerated().map { index, segment in
            """
            \(index + 1)
            \(srtTime(segment.start)) --> \(srtTime(segment.end))
            \(segment.text)
            """
        }.joined(separator: "\n\n") + "\n"
    }

    private func timecode(_ seconds: Double) -> String {
        let total = Int(seconds)
        return String(format: "%02d:%02d:%02d", total / 3600, (total % 3600) / 60, total % 60)
    }

    private func srtTime(_ seconds: Double) -> String {
        let total = Int(seconds)
        let milliseconds = Int((seconds - Double(total)) * 1000)
        return String(format: "%02d:%02d:%02d,%03d",
                      total / 3600, (total % 3600) / 60, total % 60, milliseconds)
    }
}
